apps/web/src/app/(internal)/%5F%5Fhtmltopdf/certificate/page.tsx

```diff
@@ -0,0 +1,313 @@
+import { FieldType } from '@documenso/prisma/client';
+import { Card, CardContent } from '@documenso/ui/primitives/card';
+import {
+  Table,
+  TableBody,
+  TableCell,
+  TableHead,
+  TableHeader,
+  TableRow,
+} from '@documenso/ui/primitives/table';
+
+import { Logo } from '~/components/branding/logo';
+import { LocaleDate } from '~/components/formatter/locale-date';
+
+type SigningCertificateProps = {
+  searchParams: {
+    d: string;
+  };
+};
+
+const FRIENDLY_SIGNING_REASONS = {
+  ['__OWNER__']: 'I am the owner of this document',
+  ...RECIPIENT_ROLE_SIGNING_REASONS,
+};
+
+export default async function SigningCertificate({ searchParams }: SigningCertificateProps) {
+  const { d } = searchParams;
+
+  if (typeof d !== 'string' || !d) {
+    return redirect('/');
+  }
+
+  const rawDocumentId = decryptSecondaryData(d);
+
+  if (!rawDocumentId || isNaN(Number(rawDocumentId))) {
+    return redirect('/');
+  }
+
+  const documentId = Number(rawDocumentId);
+
+  const document = await getEntireDocument({
+    id: documentId,
+  }).catch(() => null);
+
+  if (!document) {
+    return redirect('/');
+  }
+
+  const auditLogs = await getDocumentCertificateAuditLogs({
+    id: documentId,
+  });
+
+  const isOwner = (email: string) => {
+    return email.toLowerCase() === document.User.email.toLowerCase();
+  };
+
+  const getDevice = (userAgent?: string | null) => {
+    if (!userAgent) {
+      return 'Unknown';
+    }
+
+    const parser = new UAParser(userAgent);
+
+    parser.setUA(userAgent);
+
+    const result = parser.getResult();
+
+    return `${result.os.name} - ${result.browser.name} ${result.browser.version}`;
+  };
+
+  const getAuthenticationLevel = (recipientId: number) => {
+    const recipient = document.Recipient.find((recipient) => recipient.id === recipientId);
+
+    if (!recipient) {
+      return 'Unknown';
+    }
+
+    const documentAuthOptions = ZDocumentAuthOptionsSchema.parse(document.authOptions);
+    const recipientAuthOptions = ZRecipientAuthOptionsSchema.parse(recipient.authOptions);
+
+    let authLevel = 'Email';
+
+    if (
+      documentAuthOptions.globalAccessAuth === 'ACCOUNT' ||
+      recipientAuthOptions.accessAuth === 'ACCOUNT'
+    ) {
+      authLevel = 'Account Authentication';
+    }
+
+    if (
+      documentAuthOptions.globalActionAuth === 'ACCOUNT' ||
+      recipientAuthOptions.actionAuth === 'ACCOUNT'
+    ) {
+      authLevel = 'Account Re-Authentication';
+    }
+
+    if (
+      documentAuthOptions.globalActionAuth === 'TWO_FACTOR_AUTH' ||
+      recipientAuthOptions.actionAuth === 'TWO_FACTOR_AUTH'
+    ) {
+      authLevel = 'Two Factor Re-Authentication';
+    }
+
+    if (
+      documentAuthOptions.globalActionAuth === 'PASSKEY' ||
+      recipientAuthOptions.actionAuth === 'PASSKEY'
+    ) {
+      authLevel = 'Passkey Re-Authentication';
+    }
+
+    if (recipientAuthOptions.actionAuth === 'EXPLICIT_NONE') {
+      authLevel = 'Email';
+    }
```

Comment:
This seems like this might be prone to bugs where different globalAccessAuth vs accessAuth would display the wrong auth used

Example a recipient has ACCOUNT auth set, but the global is PASSKEY

Seems like it would choose PASSKEY instead of the actual method used, which is ACCOUNT

Can you use the `extractDocumentAuthMethods` util function which handles extracting the auth expected?

Reply:
I'll change that, did a minor look for some code that would do that but didn't find it initially so I did this monstrosity

Reply:
Fair, just the path name looks a little cooked %5F%5F